xbmc/video/VideoDatabase.cpp

```diff
@@ -3950,44 +3939,73 @@ void CVideoDatabase::GetSameVideoItems(CFileItem& item, CFileItemList& items)
   {
     // get items with same unique ids
     m_pDS->query(
-        PrepareSQL("SELECT media_id "
+        PrepareSQL("SELECT DISTINCT media_id "
                    "FROM uniqueid "
                    "WHERE (value, type) IN "
                    "  (SELECT value, type FROM uniqueid WHERE media_id = %i AND media_type = '%s')",
                    dbId, mediaType.c_str()));
 
     while (!m_pDS->eof())
     {
-      insertItemId(m_pDS->fv("media_id").get_asInt());
+      itemsFromUniqueIds.push_back(m_pDS->fv("media_id").get_asInt());
       m_pDS->next();
     }
 
     m_pDS->close();
 
-    // get items with same title
+    VideoDbContentType itemType = item.GetVideoContentType();
+
+    // get video item details
+    for (const auto id : itemsFromUniqueIds)
+    {
+      auto current = std::make_shared<CFileItem>();
+      if (GetDetailsByTypeAndId(*current.get(), itemType, id))
+        items.Add(current);
+    }
+
+    // get items with same title and year
+    std::vector<int> itemsFromTitleYear;
+
     std::string title = item.GetVideoInfoTag()->GetTitle();
+    int year = item.GetVideoInfoTag()->GetYear();
+    bool hasYear = item.GetVideoInfoTag()->HasYear();
 
-    VideoDbContentType itemType = item.GetVideoContentType();
     if (itemType == VideoDbContentType::MOVIES)
     {
       m_pDS->query(PrepareSQL("SELECT idMovie FROM movie WHERE movie.c%02d = '%s'",
                               VIDEODB_ID_TITLE, title.c_str()));
 
       while (!m_pDS->eof())
       {
-        insertItemId(m_pDS->fv("idMovie").get_asInt());
+        itemsFromTitleYear.push_back(m_pDS->fv("idMovie").get_asInt());
         m_pDS->next();
       }
 
       m_pDS->close();
     }
 
-    // get video item details
-    for (const auto id : itemIdList)
+    // remove the processed ids
+    itemsFromTitleYear.erase(std::remove_if(itemsFromTitleYear.begin(), itemsFromTitleYear.end(),
+                                            [&](int elem)
+                                            {
+                                              return std::find(itemsFromUniqueIds.begin(),
+                                                               itemsFromUniqueIds.end(),
+                                                               elem) != itemsFromUniqueIds.end();
+                                            }),
+                             itemsFromTitleYear.end());
```

Comment:
Probably more readable if integrated in the cursor loop just above, to add to the vector only when id isn't known yet. Would remove one level of iterator and save a few cycles.

Reply:
I'll try and explain better then!

The uniqueid query uses a media_id to select some values from the uniqueid table.  It then uses those values to select a media_id from the same table.  Because it all operates on the same table, the resultant media_id MUST always be the same as the one we supplied in the first place.

The query operates by saying "give me all the imdb & tmdb id's for the movie with id 'X'"  It then says, "Which movie has these imdb and tmdb id's?"  The answer is obviously movie 'X' every time.

Actual output from a live db

|        7900 |     1865 | movie      | 504172     | tmdb                    |
|        7901 |     1866 | movie      | tt1211837  | imdb                    |
|        7902 |     1866 | movie      | 284052     | tmdb                    |
|        7903 |     1867 | movie      | tt1764651  | imdb                    |
|        7904 |     1867 | movie      | 76163      | tmdb                    |
|        7905 |     1868 | movie      | tt0332379  | imdb                    |
|        7906 |     1868 | movie      | 1584       | tmdb                    |
|        7907 |     1869 | movie      | tt0079574  | imdb                    |
|        7908 |     1869 | movie      | 698        | tmdb                    |
|        7909 |     1870 | movie      | tt0097742  | imdb                    |
|        7910 |     1870 | movie      | 709        | tmdb                    |
|        7911 |     1871 | movie      | tt1727824  | imdb                    |
|        7912 |     1871 | movie      | 424694     | tmdb                    |
|        7913 |     1872 | movie      | tt0120737  | imdb                    |
|        7914 |     1872 | movie      | 120        | tmdb                    |
|        7915 |     1873 | movie      | tt0082398  | imdb                    |
|        7916 |     1873 | movie      | 699        | tmdb                    |
|        7917 |     1874 | movie      | tt2247732  | imdb                    |
|        7918 |     1874 | movie      | 119569     | tmdb                    |
+-------------+----------+------------+------------+-------------------------+
2370 rows in set (0.001 sec)

MariaDB [MyVideos122]> SELECT DISTINCT media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 1867 AND media_type = 'movie');
+----------+
| media_id |
+----------+
|     1867 |
+----------+
1 row in set (0.005 sec)

MariaDB [MyVideos122]> SELECT DISTINCT media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 1871 AND media_type = 'movie');
+----------+
| media_id |
+----------+
|     1871 |
+----------+
1 row in set (0.005 sec)

MariaDB [MyVideos122]> SELECT DISTINCT media_id FROM uniqueid WHERE (value, type) IN (SELECT value, type FROM uniqueid WHERE media_id = 1873 AND media_type = 'movie');
+----------+
| media_id |
+----------+
|     1873 |
+----------+
1 row in set (0.005 sec)

Reply:
my original comment was missed I think, the uniqueid table was not the point.